aws/rust-runtime/aws-runtime/src/auth.rs

```diff
@@ -178,12 +231,75 @@ pub mod sigv4 {
                 .secret_key(credentials.secret_access_key())
-                .region(operation_config.region.as_ref())
-                .service_name(operation_config.service.as_ref())
+                .region(
+                    operation_config
+                        .region
+                        .as_ref()
+                        .ok_or(SigV4SigningError::MissingSigningRegion)?
+                        .as_ref(),
+                )
+                .service_name(
+                    operation_config
+                        .service
+                        .as_ref()
+                        .ok_or(SigV4SigningError::MissingSigningService)?
+                        .as_ref(),
+                )
                 .time(request_timestamp)
                 .settings(settings);
             builder.set_security_token(credentials.session_token());
-            builder.build().expect("all required fields set")
+            Ok(builder.build().expect("all required fields set"))
+        }
+
+        fn extract_operation_config<'a>(
+            auth_scheme_endpoint_config: AuthSchemeEndpointConfig<'a>,
+            config_bag: &'a ConfigBag,
+        ) -> Result<Cow<'a, SigV4OperationSigningConfig>, SigV4SigningError> {
+            let operation_config = config_bag
+                .get::<SigV4OperationSigningConfig>()
+                .ok_or(SigV4SigningError::MissingOperationSigningConfig)?;
+
+            let EndpointAuthSchemeConfig {
+                signing_region_override,
+                signing_service_override,
+            } = Self::extract_endpoint_auth_scheme_config(auth_scheme_endpoint_config)?;
+
+            match (signing_region_override, signing_service_override) {
+                (None, None) => Ok(Cow::Borrowed(operation_config)),
+                (region, service) => {
+                    let mut operation_config = operation_config.clone();
+                    if region.is_some() {
+                        operation_config.region = region;
+                    }
+                    if service.is_some() {
+                        operation_config.service = service;
+                    }
+                    Ok(Cow::Owned(operation_config))
+                }
+            }
+        }
+
+        fn extract_endpoint_auth_scheme_config(
+            endpoint_config: AuthSchemeEndpointConfig<'_>,
+        ) -> Result<EndpointAuthSchemeConfig, SigV4SigningError> {
+            let (mut signing_region_override, mut signing_service_override) = (None, None);
+            if let Some(config) = endpoint_config.config().and_then(Document::as_object) {
+                use SigV4SigningError::BadTypeInEndpointAuthSchemeConfig as UnexpectedType;
+                signing_region_override = match config.get("signingRegion") {
+                    Some(Document::String(s)) => Some(SigningRegion::from(Region::new(s.clone()))),
+                    None => None,
+                    _ => return Err(UnexpectedType("signingRegion")),
```

Comment:
It's weird to see a `_` case for this. Is `config.get` not returning an `Option`?

Reply:
It's because I'm matching on the inner part of `Some()`. So this catches cases such as `Some(Document::Number(_))`.